### PR TITLE
A3 ides 528 web user interface v2

### PR DESCRIPTION
**Pull request by joshymjose**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -73,6 +73,14 @@ set(CUSTOM_COMPILE_OPTIONS
...
     CACHE STRING "Allows adding custom C/C++ flags"
     )
+set(WUI
+    "YES"
+    CACHE BOOL "Enable Web User Interface"
+    )
+set(GUI
+    "YES"
+    CACHE BOOL "Enable GUI"
+    )
```
Comment (mkbel) [inline CMakeLists.txt]: This will break Private. Why not use PRINTERS_WITH_GUI list?

File: CMakeLists.txt [hunk 2]
```diff
@@ -73,6 +73,14 @@ set(CUSTOM_COMPILE_OPTIONS
     ""
     CACHE STRING "Allows adding custom C/C++ flags"
     )
+set(WUI
+    "YES"
+    CACHE BOOL "Enable Web User Interface"
+    )
```
Comment (mkbel) [inline CMakeLists.txt]: Why not follow pattern PRINTERS_WITH_... list?

File: CMakeLists.txt [hunk 3]
```diff
@@ -108,20 +118,13 @@ set(unused "${CMAKE_VERBOSE_MAKEFILE} ${CMAKE_RULE_MESSAGES}")
...
   set(FILAMENT_SENSOR YES)
 else()
   set(FILAMENT_SENSOR NO)
 endif()
 
-if(${PRINTER} IN_LIST PRINTERS_WITH_GUI)
-  set(GUI YES)
-else()
-  set(GUI NO)
-endif()
```
Comment (mkbel) [inline CMakeLists.txt]: This will break private.

File: CMakeLists.txt [hunk 4]
```diff
@@ -427,13 +441,15 @@ target_sources(
...
+          src/gui/guimain.c
+          src/gui/filament.cpp
   )
 
 if(GUI)
   target_sources(
     firmware
     PRIVATE src/gui/screen_menu_filament.c
-            src/gui/dialogs/window_dlg_change.c
+    PRIVATE src/gui/dialogs/window_dlg_change.c
```
Comment (mkbel) [inline CMakeLists.txt]: ???

File: CMakeLists.txt [hunk 5]
```diff
@@ -427,13 +441,15 @@ target_sources(
           src/marlin_stubs/pause/M603.cpp
           src/marlin_stubs/pause/M701_M702.cpp
           src/marlin_stubs/pause/pause.cpp
+          src/gui/guimain.c
+          src/gui/filament.cpp
```
Comment (mkbel) [inline CMakeLists.txt]: Are you sure it won't break printers without gui?

File: CMakeLists.txt [hunk 6]
```diff
@@ -427,13 +441,15 @@ target_sources(
           src/marlin_stubs/pause/M603.cpp
           src/marlin_stubs/pause/M701_M702.cpp
           src/marlin_stubs/pause/pause.cpp
+          src/gui/guimain.c
```
Comment (mkbel) [inline CMakeLists.txt]: Are you sure it won't break printer without gui?

File: CMakeLists.txt (hunk 1 above)
Comment (joshymjose) [reply CMakeLists.txt]: my idea was that even MINI can be without GUI

File: CMakeLists.txt (hunk 6 above)
Comment (joshymjose) [reply CMakeLists.txt]: no. I have tested it

File: CMakeLists.txt (hunk 2 above)
Comment (joshymjose) [reply CMakeLists.txt]: when WUI is fully functional then GUI is not necessary, eg. in farms. I see it is good not to use gui even in MINI

File: CMakeLists.txt (hunk 4 above)
Comment (joshymjose) [reply CMakeLists.txt]: i will correct it

File: CMakeLists.txt (hunk 3 above)
Comment (joshymjose) [reply CMakeLists.txt]: in private the build script has to be modified, disabling GUI if needed

File: doc/wui/ReleaseNotes.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+Web User Interface V1
```
Comment (DRracer) [inline doc/wui/ReleaseNotes.md]: wasn't there a request for change of the file name ReleaseNotes.md -> release_notes.md already?

File: lib/AddMarlin.cmake [hunk 8]
```diff
@@ -157,6 +157,8 @@ add_library(
   )
 
 target_compile_features(Marlin PUBLIC cxx_std_14)
-target_include_directories(Marlin PUBLIC Marlin/Marlin/src Marlin/Marlin Marlin)
+target_include_directories(
+  Marlin PUBLIC Marlin/Marlin/src Marlin/Marlin/src/gcode/lcd Marlin/Marlin Marlin
+  )
```
Comment (DRracer) [inline lib/AddMarlin.cmake]: is adding "Marlin/Marlin/src/gcode/lcd" really correct?

File: src/common/eeprom.c [hunk 9]
```diff
@@ -48,6 +48,28 @@ const uint8_t eeprom_map_v1[] = {
     VARIANT8_UI8,
     VARIANT8_UI8,
     VARIANT8_UI8,
+    VARIANT8_UI32, // EEVAR_CONNECT_IP    X.X.X.X address encoded in uint32
```
Comment (DRracer) [inline src/common/eeprom.c]: coordinate this with the upcoming eeprom improvements from @XPila

File: src/common/marlin_client.c [hunk 10]
```diff
@@ -230,6 +230,17 @@ void marlin_gcode(const char *gcode) {
     _wait_ack_from_server(client->id);
 }
 
+void marlin_json_gcode(const char *gcode) {
+    char request[MARLIN_MAX_REQUEST];
+    marlin_client_t *client = _client_ptr();
+    if (client == 0)
+        return;
+    strcpy(request, "!g ");
+    strcat(request, gcode);
```
Comment (DRracer) [inline src/common/marlin_client.c]: strlcat(request, gcode, MARLIN_MAX_REQUEST)

File: src/common/marlin_client.c [hunk 11]
```diff
@@ -405,6 +416,28 @@ marlin_vars_t *marlin_update_vars(uint64_t msk) {
     return &(client->vars);
 }
 
+void marlin_set_printing_gcode_name(const char *src) {
+    char request[MARLIN_MAX_REQUEST];
+    marlin_client_t *client = _client_ptr();
+    if (client == 0)
+        return;
+    marlin_client_loop();
+    sprintf(request, "!setgcode %p", src);
```
Comment (DRracer) [inline src/common/marlin_client.c]: snprintf

File: src/common/marlin_client.c [hunk 12]
```diff
@@ -405,6 +416,28 @@ marlin_vars_t *marlin_update_vars(uint64_t msk) {
...
+    _wait_ack_from_server(client->id);
+}
+
+void marlin_get_printing_gcode_name(char *src) {
+    char request[MARLIN_MAX_REQUEST];
+    marlin_client_t *client = _client_ptr();
+    if (client == 0)
+        return;
+    marlin_client_loop();
+    sprintf(request, "!getgcode %p", src);
```
Comment (DRracer) [inline src/common/marlin_client.c]: snprintf

File: src/gui/screen_lan_settings.c [hunk 13]
```diff
@@ -72,16 +72,19 @@ static void _get_ip4_addrs(void) {
...
     strlcpy(ip4_addr_str, ip4addr_ntoa(&(config.lan_ip4_addr)), IP4_ADDR_STR_SIZE);
     strlcpy(ip4_msk_str, ip4addr_ntoa(&(config.lan_ip4_msk)), IP4_ADDR_STR_SIZE);
     strlcpy(ip4_gw_str, ip4addr_ntoa(&(config.lan_ip4_gw)), IP4_ADDR_STR_SIZE);
 
     if (flg) {
-        char save_hostname[LAN_HOSTNAME_MAX_LEN + 1];
-        eeprom_get_hostname(save_hostname);
-        snprintf(param_str, MAX_INI_SIZE, "[lan_ip4]\ntype=%s\nhostname=%s\naddress=%s\nmask=%s\ngateway=%s",
-            config.lan_flag & LAN_EEFLG_TYPE ? LAN_type_opt[1] : LAN_type_opt[0], save_hostname, ip4_addr_str, ip4_msk_str, ip4_gw_str);
+        char save_hostname[LAN_HOSTNAME_MAX_LEN + 1], save_sec_key[CONNECT_SEC_KEY_LEN + 1];
```
Comment (DRracer) [inline src/gui/screen_lan_settings.c]: Loading and saving an INI file inside the screen_lan_settings.c is not a good practice
A proper solution would be to create a serialization "object", that can load/save/notify about the config changes.
A UI screen has nothing to do with some file saving...

File: CMakeLists.txt (hunk 2 above)
Comment (DRracer) [reply CMakeLists.txt]: there is no such request so far, use PRINTES_WITH_* list

**Comment by michalrudolf** [comment]
> Cannot check the changes in Marlin, the github UI only shows the M73_PE.h
> Please address issues in comments

Only thing changed in M73_PE.h is addition of #include <stdint.h>

File: doc/wui/ReleaseNotes.md (hunk 7 above)
Comment (joshymjose) [reply doc/wui/ReleaseNotes.md]: somehow renamed in some merges, it will be renamed

File: src/common/eeprom.c (hunk 9 above)
Comment (joshymjose) [reply src/common/eeprom.c]: we are in contact, as these functions ready we will migrate

File: src/gui/screen_lan_settings.c (hunk 13 above)
Comment (joshymjose) [reply src/gui/screen_lan_settings.c]: how much time do we have to fix this?

File: lib/AddMarlin.cmake (hunk 8 above)
Comment (joshymjose) [reply lib/AddMarlin.cmake]: unfortunately this is used for print time etc. in GUI there is similar implementation with hard written path to header file used. here it is more cmake approach. Clean solution using Marlin API does not exist

**Pull request merged**